main/src/ca/uwaterloo/flix/api/EffectLock.scala

```diff
@@ -0,0 +1,348 @@
+    // TODO: Refactor this to callee (Main) and pass the root and Flix instance to this function instead
+    bootstrap.check(flix) match {
+      case Validation.Failure(errors) => Validation.Failure(errors)
+      case Validation.Success(root) =>
+        val outputStream = Files.newOutputStream(getEffectLockFile(path))
+        val signatures = mkEffectLockSignatures(root).mkString("\n") // TODO: make JSON
+        Validation.Success(outputStream.write(signatures.getBytes))
+    }
+  }
+
+
+  def mkEffectLockSignatures(root: TypedAst.Root): List[String] = {
+    root.defs.map {
+      // TODO: Consider types / type schemes and only reachable functions / visible functions
+      case (sym, defn) => s"$sym:${defn.spec}"
+    }.toList
+  }
+
+
+  /**
+    * Performs tree shaking on the given AST `root`.
+    */
+  def run(root: TypedAst.Root)(implicit flix: Flix): TypedAst.Root = {
+    // Entry points are always reachable.
+    val initReach: Set[ReachableSym] = root.entryPoints.get.map(ReachableSym.DefnSym.apply)
+
+    // Compute the symbols that are transitively reachable.
+    val allReachable = ParOps.parReach(initReach, visitSym(_, root))
+
+    // Filter the reachable definitions.
+    val reachableDefs = root.defs.filter {
+      case (sym, _) => allReachable.contains(ReachableSym.DefnSym(sym))
+    }
+
+    // Reassemble the AST.
+    root.copy(defs = reachableDefs)
+  }
+
+  /**
+    * Returns the symbols reachable from the given symbol `sym`.
+    *
+    * This includes three types of symbols:
+    *   - (a) The function or signature symbols in the implementation / body expression of a reachable function symbol
+    *   - (b) The trait symbol of a reachable sig symbol.
+    *   - (c) Every expression in a trait instance of a reachable trait symbol is reachable.
+    */
+  private def visitSym(sym: ReachableSym, root: TypedAst.Root): Set[ReachableSym] = sym match {
+    case ReachableSym.DefnSym(defnSym) =>
+      visitExp(root.defs(defnSym).exp)
+
+    case ReachableSym.SigSym(sigSym) =>
+      val sig = root.sigs(sigSym)
+      Set(ReachableSym.TraitSym(sig.sym.trt)) ++
+        sig.exp.map(visitExp).getOrElse(Set.empty)
+
+    case ReachableSym.TraitSym(traitSym) =>
+      root.instances(traitSym).foldLeft(Set.empty[ReachableSym]) {
+        case (acc, s) => visitExps(s.defs.map(_.exp)) ++ acc
+      }
+  }
+
+  /**
+    * Returns the function and signature symbols reachable from the given expression `exp0`.
+    */
+  private def visitExp(exp0: TypedAst.Expr): Set[ReachableSym] = exp0 match {
+    case Expr.Cst(_, _, _) =>
+      Set.empty
+
+    case Expr.Var(_, _, _) =>
+      Set.empty
+
+    case Expr.Hole(_, _, _, _) =>
+      Set.empty
+
+    case Expr.HoleWithExp(exp, _, _, _) =>
+      visitExp(exp)
+
+    case Expr.OpenAs(_, exp, _, _) =>
+      visitExp(exp)
+
+    case Expr.Use(sym, alias, exp, _) => // TODO: look up functions in alias
+      visitExp(exp)
+
```

Comment:
@magnus-madsen does `Resolver` update aliases in `use` or why do we not have to consider aliases?